# Segmenting and Clustering Neighborhoods in Toronto

Author: Bhargav Sagiraju

This notebook contains Questions 1, 2 & 3 of the Assignment.

## Question 1

## Importing Data

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_page = requests.get(url)
wiki_page

<Response [200]>

Response 200 means we can connect to the page

In [3]:
wiki_raw = pd.read_html(wiki_page.text)
len(wiki_raw)

3

### Creating a soup object using the Beautiful Soup library

In [4]:
soup=BeautifulSoup(wiki_page.text, 'html5lib')

### We will extract the data and create a dataframe

In [5]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['Postal Code'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

### Checking Neighborhoods which are 'Not assigned'

In [6]:
df.Neighborhood.str.count("Not assigned").sum()

0

### The resulting dataframe looks like this

In [7]:
df

Postal Code                Borough  \
0           M3A             North York   
1           M4A             North York   
2           M5A       Downtown Toronto   
3           M6A             North York   
4           M7A           Queen's Park   
..          ...                    ...   
98          M8X              Etobicoke   
99          M4Y       Downtown Toronto   
100         M7Y  East Toronto Business   
101         M8Y              Etobicoke   
102         M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df.shape

(103, 3)

### Here we see that the dataframe has 103 rows and 3 columns

## Question 2

## Installing geocoder

In [9]:
!pip install geocoder

In [10]:
import geocoder

In [11]:
# initialize your variable to None
lat_lng_coords = None

postal_code = 'M3A'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

KeyboardInterrupt: 

### Given the package is unreliable, the CSV as suggested by the assignment is used

In [12]:
data=pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [13]:
print("Wiki data shape:", df.shape)
print("CSV data shape:", data.shape)

Wiki data shape: (103, 3)
CSV data shape: (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [14]:
df.dtypes

Postal Code     object
Borough         object
Neighborhood    object
dtype: object

In [15]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code                Borough  \
0           M3A             North York   
1           M4A             North York   
2           M5A       Downtown Toronto   
3           M6A             North York   
4           M7A           Queen's Park   
..          ...                    ...   
98          M8X              Etobicoke   
99          M4Y       Downtown Toronto   
100         M7Y  East Toronto Business   
101         M8Y              Etobicoke   
102         M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [17]:
combined_data.shape

(103, 5)

Solution: We get 103 rows as expected when we do a inner join, so we have good data.

## Question 3

We cluster Toronto based on the similarities of the venues categories using Kmeans clustering and Foursquare API.

In [18]:
from geopy.geocoders import Nominatim

In [19]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


Let's visualize the map of Toronto

In [20]:
import folium

In [21]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Initializing Foursquare API credentials

In [22]:
CLIENT_ID = 'YH0KN4YRWINARM3GJXK55B15FOOWPDBT0V312BX0DHLRQWCE' 
CLIENT_SECRET = 'JUIYHSAE3X4WKW1YB5LERPFFQYGWR41CCRBHYWGVJIC5XQNB'
VERSION = '20180605' # Foursquare API version

print('Your credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentials:
CLIENT_ID: YH0KN4YRWINARM3GJXK55B15FOOWPDBT0V312BX0DHLRQWCE
CLIENT_SECRET:JUIYHSAE3X4WKW1YB5LERPFFQYGWR41CCRBHYWGVJIC5XQNB


Next, we create a function to get all the venue categories in Toronto

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting the venues in Toronto for each Neighbourhood

In [27]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [28]:
venues_in_toronto.shape

(1327, 5)

In [29]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1          TTC stop #8380               Bus Stop  
2           Variety Store      Food & Drink Shop  
3  Victoria Village Arena           Hockey Arena  
4               Portugril  Portuguese Restaurant

In [30]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                             Parkwoods   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1315  Mimico NW, The Queensway West, South of Bloor,...   
1316  Mimico NW, The Queensway West, South of Bloor,...   
1317  Mimico NW, The Queensway West, South of Bloor,...   
1318  Mimico NW, The Queensway West, South of Bloor,...   
1319  Mimico NW, The Queensway West, South of Bloor,...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.753259              -79.329656         Brookbanks Park   
1                 43.753259              -79.329656          TTC stop #8380   
2                 43.753259              -79.329656           Variety Store   
3                 43.725882              -79.315572  Victoria Village Arena   
4                 43.725882              -79.315572               Portugril   
...                     ...                     ...                     ...   
1315              43.628841              -79.520999        South St. Burger   
1316              43.628841              -79.520999              Wingporium   
1317              43.628841              -79.520999               Dollarama   
1318              43.628841              -79.520999          Healthy Planet   
1319              43.628841              -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1                  Bus Stop  
2         Food & Drink Shop  
3              Hockey Arena  
4     Portuguese Restaurant  
...                     ...  
1315           Burger Joint  
1316            Wings Joint  
1317         Discount Store  
1318        Supplement Shop  
1319                 Bakery  

[427 rows x 5 columns]

So there are 427 records for each neighbourhood.

Checking for the maximum venue categories

In [31]:
venues_in_toronto.groupby('Venue Category').max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint                             Ontario Provincial Government   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.718518              -79.464763   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Lounge                  43.628947              -79.394420   
...                                   ...                     ...   
Warehouse Store                 43.705369              -79.349372   
Wine Bar                        43.653206              -79.400049   
Wings Joint                     43.662301              -79.389494   
Women's Store                   43.689026              -79.453512   
Yoga Studio                     43.715383              -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Adult Boutique                            Seduction  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[233 rows x 4 columns]

### One Hot encode the venue Categories

In [32]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                     0               0        0                   0   
1                     0               0        0                   0   
2                     0               0        0                   0   
3                     0               0        0                   0   
4                     0               0        0                   0   
...                 ...             ...      ...                 ...   
1322                  0               0        0                   0   
1323                  0               0        0                   0   
1324                  0               0        0                   0   
1325                  0               0        0                   0   
1326                  0               0        0                   0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1322               0                0                 0                    0   
1323               0                0                 0                    0   
1324               0                0                 0                    0   
1325               0                0                 0                    0   
1326               0                0                 0                    0   

      Antique Shop  Art Gallery  ...  Train Station  \
0                0            0  ...              0   
1                0            0  ...              0   
2                0            0  ...              0   
3                0            0  ...              0   
4                0            0  ...              0   
...            ...          ...  ...            ...   
1322             0            0  ...              0   
1323             0            0  ...              0   
1324             0            0  ...              0   
1325             0            0  ...              0   
1326             0            0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1322                              0                 0            0   
1323                              0                 0            0   
1324                              0                 0            0   
1325                              0                 0            0   
1326                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1322                      0                0         0            0   
1323                      0                0         0            0   
1324                      0            

Adding the neighbourhood to the encoded dataframe

In [33]:
toronto_venue_cat['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Yoga Studio  Accessories Store  Adult Boutique  Airport  \
0            0                  0               0        0   
1            0                  0               0        0   
2            0                  0               0        0   
3            0                  0               0        0   
4            0                  0               0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  ...  Trail  Train Station  \
0                    0             0  ...      0              0   
1                    0             0  ...      0              0   
2                    0             0  ...      0              0   
3                    0             0  ...      0              0   
4                    0             0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 233 columns]

We will group the Neighbourhoods, calculate the mean venue categories in each Neighbourhood

In [34]:
toronto_grouped = toronto_venue_cat.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Agincourt          0.0   
1                           Alderwood, Long Branch          0.0   
2  Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                  Bayview Village          0.0   
4                Bedford Park, Lawrence Manor East          0.0   

   Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                0.0             0.0      0.0                 0.0   
1                0.0             0.0      0.0                 0.0   
2                0.0             0.0      0.0                 0.0   
3                0.0             0.0      0.0                 0.0   
4                0.0             0.0      0.0                 0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.037037   

   ...  Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  ...    0.0            0.0                            0.0               0.0   
1  ...    0.0            0.0                            0.0               0.0   
2  ...    0.0            0.0                            0.0               0.0   
3  ...    0.0            0.0                            0.0               0.0   
4  ...    0.0            0.0                            0.0               0.0   

   Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0          0.0                    0.0              0.0       0.0          0.0   
1          0.0                    0.0              0.0       0.0          0.0   
2          0.0                    0.0              0.0       0.0          0.0   
3          0.0                    0.0              0.0       0.0          0.0   
4          0.0                    0.0              0.0       0.0          0.0   

   Women's Store  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  

[5 rows x 233 columns]

Let's write a function to get the top most common venue categories

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

There are way too many venue categories, let's take the top 10 to cluster the neighbourhoods

In [37]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Breakfast Spot   
1           Coffee Shop          Skating Rink                   Gym   
2           Coffee Shop           Pizza Place           Supermarket   
3                  Bank   Japanese Restaurant                  Café   
4           Coffee Shop    Italian Restaurant            Hobby Shop   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0         Deli / Bodega             Drugstore                   Donut Shop   
1                   Pub        Sandwich Place                 Dessert Shop   
2        Ice Cream Shop      Sushi Restaurant                   Restaurant   
3         Women's Store         Deli / Bodega  Eastern European Restaurant   
4      Greek Restaurant        Breakfast Spot                    Juice Bar   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0               Dog Run   Distribution Center             Discount Store  
1          Dance Studio         Deli / Bodega           Department Store  
2                 Diner         Deli / Bodega  Middle Eastern Restaurant  
3             Drugstore            Donut Shop                    Dog Run  
4          Liquor Store            Restaurant        Japanese Restaurant

Let's make the model to cluster our Neighbourhoods

In [38]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [39]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Checking the labelling of our model

In [40]:
kmeans.labels_[0:100]

array([2, 1, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 1, 2, 2, 2, 1, 1, 2,
       2, 2, 2, 3, 1, 1, 1, 2, 1, 2, 2, 2, 2, 2, 2, 1, 1, 2, 2, 2, 1, 1,
       0, 2, 1, 2, 2, 1, 2, 1, 2, 2, 1, 3, 1, 1, 1, 1, 2, 2, 0, 2, 2, 1,
       3, 2, 2, 3, 2, 2, 2, 2, 2, 1, 2, 2, 1, 2, 2, 2, 2, 3, 2, 3, 1, 2,
       2, 1, 4, 1, 3, 1, 2, 1, 3, 2, 1, 3])

Let's add the clustering Label column to the top 10 common venue categories

In [41]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Join toronto_grouped with combined_data on neighborhood to add latitude & longitude for each neighborhood to prepare it for plotting

In [42]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3         M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4         M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue   2nd Most Common Venue  \
0 -79.329656             3.0              Bus Stop                    Park   
1 -79.315572             1.0           Coffee Shop             Pizza Place   
2 -79.360636             2.0           Coffee Shop                  Bakery   
3 -79.464763             2.0        Clothing Store  Furniture / Home Store   
4 -79.389494             2.0           Coffee Shop        Sushi Restaurant   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0     Food & Drink Shop         Women's Store              Drugstore   
1          Hockey Arena          Intersection  Portuguese Restaurant   
2                  Park                  Café         Breakfast Spot   
3    Miscellaneous Shop     Accessories Store               Boutique   
4           Yoga Studio              Creperie    Distribution Center   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Donut Shop               Dog Run   Distribution Center   
1          Women's Store                 Diner      Department Store   
2            Yoga Studio  Gym / Fitness Center     French Restaurant   
3  Vietnamese Restaurant             Gift Shop           Coffee Shop   
4         Discount Store                 Diner         Burrito Place   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner  
1          Dessert Shop     Dim Sum Restaurant  
2         Historic Site         Farmers Market  
3   Arts & Crafts Store       Department Store  
4         Smoothie Shop    Japanese Restaurant

Dropping all the NaN values to prevent data skew

In [43]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plotting the clusters on the map

In [44]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [45]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighborhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Let's verify each of our clusters

Cluster 1

In [46]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             0.0        Baseball Field         Women's Store   
101   Etobicoke             0.0        Baseball Field         Women's Store   

    3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
57          Deli / Bodega     Electronics Store  Eastern European Restaurant   
101         Deli / Bodega     Electronics Store  Eastern European Restaurant   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57              Drugstore            Donut Shop               Dog Run   
101             Drugstore            Donut Shop               Dog Run   

    9th Most Common Venue 10th Most Common Venue  
57    Distribution Center         Discount Store  
101   Distribution Center         Discount Store

Cluster 2

In [47]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1               North York             1.0                Coffee Shop   
6              Scarborough             1.0       Fast Food Restaurant   
8                East York             1.0                Pizza Place   
10              North York             1.0                       Park   
14               East York             1.0         Athletics & Sports   
27              North York             1.0                Golf Course   
29               East York             1.0          Indian Restaurant   
31            West Toronto             1.0                   Pharmacy   
44             Scarborough             1.0                     Bakery   
46              North York             1.0              Grocery Store   
47            East Toronto             1.0             Sandwich Place   
49              North York             1.0                       Park   
50              North York             1.0                Pizza Place   
56                    York             1.0       Fast Food Restaurant   
60              North York             1.0         Athletics & Sports   
61         Central Toronto             1.0                       Park   
67         Central Toronto             1.0                       Park   
70               Etobicoke             1.0                Pizza Place   
71             Scarborough             1.0  Middle Eastern Restaurant   
72              North York             1.0                   Pharmacy   
77               Etobicoke             1.0          Mobile Phone Shop   
82             Scarborough             1.0                   Pharmacy   
83         Central Toronto             1.0                 Restaurant   
88               Etobicoke             1.0                 Restaurant   
89               Etobicoke             1.0              Grocery Store   
90             Scarborough             1.0                Coffee Shop   
93               Etobicoke             1.0                Pizza Place   
100  East Toronto Business             1.0                Yoga Studio   
102              Etobicoke             1.0              Tanning Salon   

    2nd Most Common Venue       3rd Most Common Venue  \
1             Pizza Place                Hockey Arena   
6           Women's Store                Dance Studio   
8          Breakfast Spot                 Flea Market   
10            Pizza Place         Japanese Restaurant   
14                   Park                  Beer Store   
27                Dog Run        Fast Food Restaurant   
29            Yoga Studio                 Coffee Shop   
31                 Bakery                        Bank   
44               Bus Line               Metro Station   
46                   Park               Shopping Mall   
47            Pizza Place        Fast Food Restaurant   
49       Basketball Court  Construction & Landscaping   
50                    Gym                Intersection   
56         Discount Store              Sandwich Place   
60          Grocery Store                Liquor Store   
61               Bus Line                 Swim School   
67                  Hotel              Breakfast Spot   
70            Coffee Shop              Discount Store   
71            Auto Garage                  Smoke Shop   
72            Supermarket                 Pizza Place   
77                   Park              Sandwich Place   
82            Pizza Place                Noodle House   
83                   Park                Tennis Court   
88                   Café                Liquor Store   
89         Sandwich Place                    Pharmacy   
90   Fast Food Restaurant              Sandwich Place   
93            Coffee Shop                Skating Rink   
100            Comic Shop                     Brewery   
102        Sandwich Place             Supplement Shop   

           4th Most Common Venue     5th Most Common Venue  \
1                   Intersection     Portuguese Restaur

Cluster 3

In [48]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
2         Downtown Toronto             2.0                Coffee Shop   
3               North York             2.0             Clothing Store   
4             Queen's Park             2.0                Coffee Shop   
7               North York             2.0                        Gym   
9         Downtown Toronto             2.0                       Café   
12             Scarborough             2.0                        Bar   
13              North York             2.0                 Restaurant   
15        Downtown Toronto             2.0                  Gastropub   
16                    York             2.0                      Field   
17               Etobicoke             2.0                  Pet Store   
18             Scarborough             2.0          Electronics Store   
19            East Toronto             2.0                      Trail   
20        Downtown Toronto             2.0               Cocktail Bar   
22             Scarborough             2.0                Coffee Shop   
23               East York             2.0                Coffee Shop   
24        Downtown Toronto             2.0                Coffee Shop   
25        Downtown Toronto             2.0              Grocery Store   
26             Scarborough             2.0                     Bakery   
28              North York             2.0                       Bank   
30        Downtown Toronto             2.0                Coffee Shop   
32             Scarborough             2.0                 Playground   
33              North York             2.0             Clothing Store   
34              North York             2.0         Falafel Restaurant   
36        Downtown Toronto             2.0                       Park   
37            West Toronto             2.0                        Bar   
38             Scarborough             2.0         Chinese Restaurant   
39              North York             2.0         Chinese Restaurant   
41            East Toronto             2.0           Greek Restaurant   
42        Downtown Toronto             2.0                       Café   
43            West Toronto             2.0                       Café   
48        Downtown Toronto             2.0                       Café   
51             Scarborough             2.0               Skating Rink   
53              North York             2.0           Business Service   
54            East Toronto             2.0                Coffee Shop   
55              North York             2.0             Sandwich Place   
58             Scarborough             2.0            College Stadium   
59              North York             2.0                Coffee Shop   
62         Central Toronto             2.0    Health & Beauty Service   
63                    York             2.0             Breakfast Spot   
65             Scarborough             2.0          Indian Restaurant   
68         Central Toronto             2.0              Jewelry Store   
69            West Toronto             2.0         Mexican Restaurant   
73         Central Toronto             2.0             Clothing Store   
74         Central Toronto             2.0             Sandwich Place   
75            West Toronto             2.0                  Gift Shop   
76             Mississauga             2.0                Coffee Shop   
78             Scarborough             2.0  Latin American Restaurant   
79         Central Toronto             2.0               Dessert Shop   
80        Downtown Toronto             2.0                       Café   
81            West Toronto             2.0                       Café   
84        Downtown Toronto             2.0                       Café   
86         Central Toronto             2.0                Coffee Shop   
87        Downtown Toronto             2.0            Airport Service   
92  Downtown Toronto Stn A             2.0         Seafood Restaurant   
94  

Cluster 4

In [49]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             3.0              Bus Stop   
21                    York             3.0                  Park   
35  East York/East Toronto             3.0                  Park   
40              North York             3.0      Business Service   
52              North York             3.0                  Park   
64                    York             3.0                  Park   
66              North York             3.0                  Park   
85             Scarborough             3.0                  Park   
91        Downtown Toronto             3.0                  Park   
98               Etobicoke             3.0                 River   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                   Park     Food & Drink Shop                Women's Store   
21         Women's Store                  Pool                  Curling Ice   
35     Convenience Store         Women's Store                 Dance Studio   
40                  Park               Airport                Women's Store   
52         Women's Store           Curling Ice  Eastern European Restaurant   
64     Convenience Store         Women's Store                 Dance Studio   
66     Convenience Store         Women's Store                 Dance Studio   
85            Playground          Intersection                Women's Store   
91            Playground                 Trail             Cuban Restaurant   
98                  Park         Women's Store                    Drugstore   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21                    Drugstore            Donut Shop               Dog Run   
35  Eastern European Restaurant             Drugstore            Donut Shop   
40  Eastern European Restaurant             Drugstore            Donut Shop   
52                    Drugstore            Donut Shop               Dog Run   
64  Eastern European Restaurant             Drugstore            Donut Shop   
66  Eastern European Restaurant             Drugstore            Donut Shop   
85                  Curling Ice             Drugstore            Donut Shop   
91                    Drugstore            Donut Shop               Dog Run   
98                   Donut Shop               Dog Run   Distribution Center   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Distribution Center        Discount Store                  Diner  
21   Distribution Center        Discount Store                  Diner  
35               Dog Run   Distribution Center         Discount Store  
40               Dog Run   Distribution Center         Discount Store  
52   Distribution Center        Discount Store                  Diner  
64               Dog Run   Distribution Center         Discount Store  
66               Dog Run   Distribution Center         Discount Store  
85               Dog Run   Distribution Center         Discount Store  
91   Distribution Center        Discount Store                  Diner  
98        Discount Store                 Diner     Dim Sum Restaurant

Cluster 5

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0   Filipino Restaurant         Women's Store   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
11          Dance Studio  Eastern European Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11            Donut Shop               Dog Run   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
11        Discount Store                  Diner

We have now successfully clustered Toronto neighbourhoods based on venue categories!